docs/performance-metrics.md

```diff
@@ -0,0 +1,98 @@
+
+It is perfectly normal for etcd to use a multiple of its "in use DB size" as resident memory.
+
+Try to look through the audit log and find service accounts that might be listing pods or configmaps across namespaces.
+Another frequent offender are events from chatty operators, which can be found in the audit log or directly from [etcd using KCS6970015](https://access.redhat.com/solutions/6970015).
+
+#### Defrag and Compaction
+
+One way to reduce memory usage is to defrag bbolt and compact the database to the latest revision. It is important to run the defrag after the compaction, so one would execute:
+
+> etctctl compact $(etctctl get / -w json | jq .header.revision)
+> 
+> etctctl defrag
+
+In more recent OpenShift versions this is done automatically -  compacting from the API server and defrag from the etcd-operator.
+
+Sometimes etcd is quite reluctant to free memory, due to the laziness of the Golang GC. So it is perfectly normal for the resident memory to stay up until memory pressure forces it to go down again.
+
+### CPU
+
+While ETCD is not very heavy user of CPU in general, similar to memory, ETCD is competing for CPU usage with other components on the same hosts.
+
+This query will give you our most offending candidates (API server, OVN and ETCD) in a single chart. 
+
+> sum(node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate{namespace=~"openshift-etcd|openshift-ovn-kubernetes|openshift-apiserver"}) by (namespace)
+
+Especially when dealing with OVN, you might want to drill deeper into what component is using the CPU:
+
+> rate(container_cpu_usage_seconds_total{namespace=~"openshift-etcd|openshift-ovn-kubernetes", container=~"etcd|ovnkube-master|nbdb|northd|sbdb"}[2m])
+
+Keep in mind that this is not exhaustive, there might be something else that runs on the host that consumes lots of CPU and thus starves etcd. It also makes sense to look at the whole picture with:
+
+> sum(node_namespace_pod_container:container_cpu_usage_seconds_total:sum_irate{cluster="", node=~"!!! NODE NAME !!!"}) by (pod)
+ 
+
+### Disk Latency
+
+ETCD is extremely sensitive to high latency disk latency. ETCD utilizes the disk to persist bbolt's memory mapped file, but more importantly, for writing to the write-ahead-log (WAL). The latter is the most important path with writing and executes fsync to ensure the data is durably synchronized to the disk.
+
+From ETCD, you can look at the 99th percentile of the WAL append with:
+
+> histogram_quantile(0.99, sum(rate(etcd_disk_wal_fsync_duration_seconds_bucket{job="etcd"}[5m])) by (instance, le))
+
+To ensure the best performance, this metric should not exceed 10ms. It's thus incredibly important to run OpenShift ETCD on SSD or better NVME drives that offer less than a 1ms.
+
+The backend disk commit can be tracked on 99th percentile with: 
+
+> histogram_quantile(0.99, sum(rate(etcd_disk_backend_commit_duration_seconds_bucket{job="etcd"}[5m])) by (instance, le))
```

Comment:
My PromQL-fu is severely lacking 😢  so this might be a stupid question but is this the same as other variants I've seen documented elsewhere:

- The [upstream alert query](https://github.com/openshift/cluster-etcd-operator/blob/33fdaf64825c13f578254107f7e7f0d503c724c9/jsonnet/vendor/github.com/etcd-io/etcd/contrib/mixin/mixin.libsonnet#L201) just uses rate and does not sum.
  - `histogram_quantile(0.99, rate(etcd_disk_backend_commit_duration_seconds_bucket{}[5m]))`
- KCS article uses irate: https://access.redhat.com/solutions/5489721
  - `histogram_quantile(0.99, irate(etcd_disk_backend_commit_duration_seconds_bucket[5m]))`
  
Although I have this exact query documented somewhere in my notes from when I saw Sam using it so I'm fairly certain this is what we've always used.

Reply:
great collection :+1: 

irate handles counter resets properly, so I'll use that instead - thanks!
sum is indeed not needed in our case, not sure why I added it in here... also fixed